File: src/tak/board.rs
use std::collections::VecDeque;

use crate::tak::{TakDir, TakPlayer, TakPos, TakVariant};

#[derive(Clone, Debug)]
pub struct TakStack {
    variant: TakVariant,
    composition: Vec<TakPlayer>,
}

#[derive(Clone, Debug)]
pub struct TakBoard {
    size: u32,
    stacks: Vec<Option<TakStack>>,
}

impl TakBoard {
    pub fn new(size: u32) -> Self {
        let board_area = size * size;
        TakBoard {
            size,
            stacks: vec![None; board_area as usize],
        }
    }

    pub fn can_do_place(&self, pos: &TakPos) -> Result<(), String> {
        if !pos.is_valid(self.size) {
            return Err("Position out of bounds".to_string());
        }
        let index = (pos.y * self.size as i32 + pos.x) as usize;
        if self.stacks[index].is_some() {
            return Err("Position already occupied".to_string());
        }
        Ok(())
    }

    pub fn do_place(
        &mut self,
        pos: &TakPos,
        variant: &TakVariant,
        player: &TakPlayer,
    ) -> Result<(), String> {
        self.can_do_place(&pos)?;
        let index = (pos.y * self.size as i32 + pos.x) as usize;
        self.stacks[index] = Some(TakStack {
            variant: variant.clone(),
            composition: vec![player.clone()],
        });
        Ok(())
    }

    pub fn can_do_move(&self, pos: &TakPos, dir: &TakDir, drops: &[u32]) -> Result<(), String> {
        if !pos.is_valid(self.size) {
            return Err("Position out of bounds".to_string());
        }
        let index = (pos.y * self.size as i32 + pos.x) as usize;
        let stack = self.stacks[index]
            .as_ref()
            .ok_or_else(|| "No stack at the given position".to_string())?;
        let total_pieces: u32 = stack.composition.len() as u32;
        let drops_sum: u32 = drops.iter().sum();
        if drops_sum == 0 || drops_sum > total_pieces || drops_sum > self.size {
            return Err("Invalid number of pieces to move".to_string());
        }
        let drops_len = drops.len();
        let end_pos = pos.offset(dir, drops_len as i32);
        if !end_pos.is_valid(self.size) {
            return Err("Move goes out of bounds".to_string());
        }
        for i in 0..drops_len {
            if drops[i] == 0 {
                return Err("Drop count cannot be zero".to_string());
            }
            let cur_pos = pos.offset(dir, i as i32 + 1);
            let cur_index = (cur_pos.y * self.size as i32 + cur_pos.x) as usize;

            match &self.stacks[cur_index] {
                Some(TakStack {
                    variant: TakVariant::Standing,
                    composition: _,
                }) if stack.variant != TakVariant::Capstone
                    || i < drops_len - 1
                    || drops[i] != 1 =>
                {
                    return Err("Cannot move over standing pieces".to_string());
                }
                Some(TakStack {
                    variant: TakVariant::Capstone,
                    composition: _,
                }) => {
                    return Err("Cannot move over capstone pieces".to_string());
                }
                _ => {}
            }
        }
        Ok(())
    }

    pub fn do_move(&mut self, pos: &TakPos, dir: &TakDir, drops: &[u32]) -> Result<(), String> {
        self.can_do_move(pos, dir, drops)?;
        let index = (pos.y * self.size as i32 + pos.x) as usize;
        let stack = self.stacks[index].as_mut().unwrap();
        let total_pieces: u32 = stack.composition.len() as u32;
        let drops_sum: u32 = drops.iter().sum();
        let variant = stack.variant.clone();
        let mut moving_pieces = if drops_sum == total_pieces {
            self.stacks[index].take().unwrap().composition
        } else {
            stack
                .composition
                .split_off((total_pieces - drops_sum) as usize)
        };
        moving_pieces.reverse();

        let drops_len = drops.len();
        for i in 0..drops_len {
            let current_pos = pos.offset(dir, i as i32 + 1);
            let cur_index = (current_pos.y * self.size as i32 + current_pos.x) as usize;
            let cur_stack = self.stacks[cur_index].get_or_insert(TakStack {
                variant: TakVariant::Flat,
                composition: vec![],
            });
            let to_drop = moving_pieces
                .drain(moving_pieces.len() - drops[i] as usize..)
                .rev();
            cur_stack.composition.extend(to_drop);
            if i == drops_len - 1 {
                cur_stack.variant = variant.clone();
            }
        }
        Ok(())
    }

    pub fn is_full(&self) -> bool {
        self.stacks.iter().all(|s| s.is_some())
    }

    pub fn count_flats(&self) -> (u32, u32) {
        let mut white_flats = 0;
        let mut black_flats = 0;
        for stack in &self.stacks {
            if let Some(s) = stack {
                if let Some(top_player) = s.composition.last() {
                    match s.variant {
                        TakVariant::Flat => match top_player {
                            TakPlayer::White => white_flats += 1,
                            TakPlayer::Black => black_flats += 1,
                        },
                        _ => {}
                    }
                }
            }
        }
        (white_flats, black_flats)
    }

    fn is_road_square(&self, pos: &TakPos, player: &TakPlayer) -> bool {
        if !pos.is_valid(self.size) {
            return false;
        }
        let index = (pos.y * self.size as i32 + pos.x) as usize;
        if let Some(stack) = &self.stacks[index] {
            if let Some(top_player) = stack.composition.last() {
                return stack.variant != TakVariant::Standing && top_player == player;
            }
        }
        false
    }

    pub fn check_for_road(&self, player: &TakPlayer) -> bool {
        self.find_road(true, player) || self.find_road(false, player)
    }

    fn find_road(&self, horizontal: bool, player: &TakPlayer) -> bool {
        let mut visited = vec![false; (self.size * self.size) as usize];
        let mut queue = VecDeque::new();

        for i in 0..self.size as i32 {
            let start_pos = if horizontal {
                TakPos { x: 0, y: i }
            } else {
                TakPos { x: i, y: 0 }
            };
            if self.is_road_square(&start_pos, player) {
                queue.push_back(start_pos.clone());
                let index = (start_pos.y * self.size as i32 + start_pos.x) as usize;
                visited[index] = true;
            }
        }

        while let Some(current_pos) = queue.pop_front() {
            let is_end = if horizontal {
                current_pos.x == self.size as i32 - 1
            } else {
                current_pos.y == self.size as i32 - 1
            };
            if is_end {
                return true;
            }

            for dir in &[TakDir::Up, TakDir::Down, TakDir::Left, TakDir::Right] {
                let neighbor = current_pos.offset(dir, 1);
                if self.is_road_square(&neighbor, player) {
                    let index = (neighbor.y * self.size as i32 + neighbor.x) as usize;
                    if !visited[index] {
                        queue.push_back(neighbor.clone());
                        visited[index] = true;
                    }
                }
            }
        }

        false
    }

    pub fn compute_hash_string(&self) -> String {
        let mut hash = Vec::new();
        for stack in &self.stacks {
            match stack {
                Some(s) => {
                    let variant_char = match s.variant {
                        TakVariant::Flat => 'F',
                        TakVariant::Standing => 'S',
                        TakVariant::Capstone => 'C',
                    };
                    let composition_str: String = s
                        .composition
                        .iter()
                        .map(|p| match p {
                            TakPlayer::White => 'W',
                            TakPlayer::Black => 'B',
                        })
                        .collect();
                    hash.push(format!("{}{}", variant_char, composition_str));
                }
                None => {
                    hash.push("N".to_string());
                }
            }
        }
        hash.join(",")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_place_and_move() {
        let mut board = TakBoard::new(5);
        let pos = TakPos { x: 0, y: 0 };
        let player = TakPlayer::White;
        let variant = TakVariant::Flat;

        assert!(board.do_place(&pos, &variant, &player).is_ok());
        assert!(board.can_do_move(&pos, &TakDir::Right, &[1]).is_ok());
        assert!(board.do_move(&pos, &TakDir::Right, &[1]).is_ok());

        let new_pos = TakPos { x: 1, y: 0 };
        assert!(board.stacks[(new_pos.y * board.size as i32 + new_pos.x) as usize].is_some());
    }

    #[test]
    fn test_block() {
        let mut board = TakBoard::new(5);
        let pos = TakPos { x: 1, y: 1 };
        let wall_pos = TakPos { x: 2, y: 1 };
        let cap_pos = TakPos { x: 1, y: 2 };
        let player = TakPlayer::White;
        let variant = TakVariant::Flat;

        assert!(board.do_place(&pos, &variant, &player).is_ok());
        assert!(
            board
                .do_place(&wall_pos, &TakVariant::Standing, &player)
                .is_ok()
        );
        assert!(
            board
                .do_place(&cap_pos, &TakVariant::Capstone, &player)
                .is_ok()
        );
        assert!(board.can_do_move(&pos, &TakDir::Right, &[1]).is_err());
        assert!(board.can_do_move(&pos, &TakDir::Up, &[1]).is_err());
    }

    #[test]
    fn test_caps_smash() {
        let mut board = TakBoard::new(5);
        let pos = TakPos { x: 0, y: 0 };
        let cap_pos = TakPos { x: 1, y: 0 };
        let player = TakPlayer::White;
        let variant = TakVariant::Capstone;

        assert!(board.do_place(&pos, &variant, &player).is_ok());
        assert!(
            board
                .do_place(&cap_pos, &TakVariant::Standing, &player)
                .is_ok()
        );
        assert_eq!(board.can_do_move(&pos, &TakDir::Right, &[1]), Ok(()));
        assert!(board.do_move(&pos, &TakDir::Right, &[1]).is_ok());
    }

    #[test]
    fn test_road_detection() {
        let mut board = TakBoard::new(5);
        let player = TakPlayer::White;
        let variant = TakVariant::Flat;

        for x in 0..5 {
            let pos = TakPos { x, y: 0 };
            assert!(board.do_place(&pos, &variant, &player).is_ok());
        }

        assert!(board.check_for_road(&player));
    }
}

File: packages/tak-frontend/src/tak-core/board.ts
import { immerable } from 'immer';
import {
  allDirections,
  isValidPos,
  offsetPos,
  type TakDir,
  type TakPieceId,
  type TakPlayer,
  type TakPos,
  type TakVariant,
} from '.';

export interface TakStack {
  variant: TakVariant;
  composition: TakPieceId[];
}

export class TakBoard {
  [immerable] = true;

  size: number;
  stacks: (TakStack | undefined)[];
  uuids: string[] = [];
  givenUuidCount: number = 0;
  private kindIndices: Record<TakPlayer, Record<'flat' | 'capstone', number>> = {
    white: { flat: 0, capstone: 0 },
    black: { flat: 0, capstone: 0 },
  };

  constructor(size: number) {
    this.size = size;
    this.stacks = Array.from({ length: size * size })
      .fill(undefined)
      .map(() => undefined);
  }

  clone(): TakBoard {
    const newBoard = new TakBoard(this.size);
    newBoard.stacks = this.stacks.map((stack) => {
      if (stack === undefined) {
        return undefined;
      }
      return {
        variant: stack.variant,
        composition: stack.composition.map((piece) => ({ ...piece })),
      };
    });
    newBoard.uuids = [...this.uuids];
    newBoard.givenUuidCount = this.givenUuidCount;
    newBoard.kindIndices = {
      white: { ...this.kindIndices.white },
      black: { ...this.kindIndices.black },
    };
    return newBoard;
  }

  addUuids(uuids: string[]) {
    this.uuids.push(...uuids);
  }

  trimExcessUuids() {
    this.uuids = this.uuids.slice(0, this.givenUuidCount);
  }

  getStack(pos: TakPos): TakStack | undefined {
    if (!isValidPos(this.size, pos)) {
      return undefined;
    }
    const index = pos.y * this.size + pos.x;
    return this.stacks[index];
  }

  canDoPlace(pos: TakPos): boolean {
    if (!isValidPos(this.size, pos)) {
      return false;
    }
    const index = pos.y * this.size + pos.x;
    return this.stacks[index] === undefined;
  }

  doPlace(pos: TakPos, variant: TakVariant, composition: TakPlayer[]): TakPieceId[] {
    if (!this.canDoPlace(pos)) {
      throw new Error(`Cannot place at position (${pos.x.toString()}, ${pos.y.toString()})`);
    }
    const index = pos.y * this.size + pos.x;
    const pieceType = variant === 'capstone' ? 'capstone' : 'flat';
    const pieceIds: TakPieceId[] = composition.map((player) => {
      const uuidIndex = this.givenUuidCount++;
      const kindIndex = this.kindIndices[player][pieceType]++;
      const uuid = this.uuids[uuidIndex] ?? crypto.randomUUID();
      this.uuids[uuidIndex] = uuid;
      return {
        uuid,
        type: pieceType,
        player,
        kindIndex,
      };
    });
    this.stacks[index] = {
      variant,
      composition: [...pieceIds],
    };
    return pieceIds;
  }

  canDoMove(pos: TakPos, dir: TakDir, drops: number[]): boolean {
    if (!isValidPos(this.size, pos)) {
      return false;
    }
    const index = pos.y * this.size + pos.x;
    const stack = this.stacks[index];
    if (stack === undefined) {
      return false;
    }
    const dropsSum = drops.reduce((a, b) => a + b, 0);
    if (dropsSum <= 0 || dropsSum > this.size || dropsSum > stack.composition.length) {
      return false;
    }
    const endPos = offsetPos(pos, dir, drops.length);
    if (!isValidPos(this.size, endPos)) {
      return false;
    }
    for (let i = 0; i < drops.length; i++) {
      const dropAmount = drops[i];
      if (dropAmount === undefined || dropAmount <= 0) {
        return false;
      }
      const curPos = offsetPos(pos, dir, i + 1);
      const curIndex = curPos.y * this.size + curPos.x;
      const curStack = this.stacks[curIndex];

      const canSmash = stack.variant === 'capstone' && i === drops.length - 1 && dropAmount === 1;
      if (
        curStack !== undefined &&
        (curStack.variant === 'capstone' || (curStack.variant === 'standing' && !canSmash))
      ) {
        return false;
      }
    }
    return true;
  }

  doMove(pos: TakPos, dir: TakDir, drops: number[]): { pieceIds: TakPieceId[]; wasSmash: boolean } {
    if (!this.canDoMove(pos, dir, drops)) {
      throw new Error(`Cannot move from position (${pos.x.toString()}, ${pos.y.toString()})`);
    }
    const index = pos.y * this.size + pos.x;
    const stack = this.stacks[index];
    if (!stack) {
      throw new Error(
        `Unreachable: stack should exist at position (${pos.x.toString()}, ${pos.y.toString()})`,
      );
    }
    const dropsSum = drops.reduce((a, b) => a + b, 0);
    const movingPieces = stack.composition.splice(-dropsSum);
    const movingPieceIds = [...movingPieces];

    const variant = stack.variant;
    stack.variant = 'flat';
    if (stack.composition.length === 0) {
      this.stacks[index] = undefined;
    }
    movingPieces.reverse();
    let wasSmash = false;

    for (let i = 0; i < drops.length; i++) {
      const curPos = offsetPos(pos, dir, i + 1);
      const curIndex = curPos.y * this.size + curPos.x;
      let curStack = this.stacks[curIndex];
      if (curStack === undefined) {
        curStack = { variant: 'flat', composition: [] };
        this.stacks[curIndex] = curStack;
      }
      const toDrop = movingPieces.splice(-(drops[i] ?? 0));
      toDrop.reverse();
      curStack.composition.push(...toDrop);
      if (i === drops.length - 1) {
        if (curStack.variant === 'standing') {
          wasSmash = true;
        }
        curStack.variant = variant;
      }
    }
    return { pieceIds: movingPieceIds, wasSmash };
  }

  isFull(): boolean {
    return this.stacks.every((stack) => stack !== undefined);
  }

  countFlats(): Record<TakPlayer, number> {
    const counts = {
      white: 0,
      black: 0,
    };
    for (const stack of this.stacks) {
      if (stack !== undefined) {
        const topPiece = stack.composition[stack.composition.length - 1];
        if (topPiece && stack.variant === 'flat') {
          counts[topPiece.player]++;
        }
      }
    }
    return counts;
  }

  private isRoadSquare(pos: TakPos, player: TakPlayer): boolean {
    if (!isValidPos(this.size, pos)) {
      return false;
    }
    const index = pos.y * this.size + pos.x;
    const stack = this.stacks[index];
    if (stack === undefined || stack.variant === 'standing') {
      return false;
    }
    const topPiece = stack.composition[stack.composition.length - 1];
    return topPiece !== undefined && topPiece.player === player;
  }

  checkForRoad(player: TakPlayer): TakPos[] | null {
    return this.findRoad(true, player) ?? this.findRoad(false, player);
  }

  private findRoad(horizontal: boolean, player: TakPlayer): TakPos[] | null {
    const visited = Array.from({ length: this.size * this.size }, () => false);
    const prev: (TakPos | null)[] = Array.from({ length: this.size * this.size }, () => null);
    const queue: TakPos[] = [];

    const toIndex = (p: TakPos) => p.y * this.size + p.x;

    for (let i = 0; i < this.size; i++) {
      const pos = horizontal ? { x: 0, y: i } : { x: i, y: 0 };
      if (this.isRoadSquare(pos, player)) {
        const idx = toIndex(pos);
        queue.push(pos);
        visited[idx] = true;
      }
    }

    while (queue.length > 0) {
      const pos = queue.shift();
      if (!pos) {
        break;
      }

      const isEnd = (horizontal ? pos.x : pos.y) === this.size - 1;
      if (isEnd) {
        const path: TakPos[] = [];
        let cur: TakPos | null = pos;

        while (cur) {
          path.push(cur);
          cur = prev[toIndex(cur)] ?? null;
        }

        path.reverse();
        return path;
      }

      for (const dir of allDirections) {
        const neighbor = offsetPos(pos, dir, 1);

        if (!this.isRoadSquare(neighbor, player)) continue;

        const nIdx = toIndex(neighbor);
        if (visited[nIdx] === true) continue;

        visited[nIdx] = true;
        prev[nIdx] = pos;
        queue.push(neighbor);
      }
    }

    return null;
  }

  computeHashString(): string {
    return this.stacks
      .map((stack) => {
        if (stack !== undefined) {
          const variantChar =
            stack.variant === 'flat' ? 'F' : stack.variant === 'standing' ? 'S' : 'C';
          const compositionStr = stack.composition
            .map((piece) => (piece.player === 'white' ? 'W' : 'B'))
            .join('');
          return variantChar + compositionStr;
        } else {
          return 'N';
        }
      })
      .join(',');
  }
}
